### ISSUE TITLE
Make the return type of`fc.object(...)` dependent on its input

### ISSUE THREAD

**Issue by TomerAberbach** [dubzzz/fast-check#3495]
## 🚀 Feature Request

Right now `fc.object(...)` is typed as always returning `Arbitrary<Record<string, unknown>>`, but if I pass arbitraries for the `key` and/or `values` options, then I would expect the return type to reflect that.

## Example

Maybe something like this? The types seem to infer correctly from my testing!

```ts
declare function object<Constraints extends ObjectConstraints>(
  constraints: Constraints,
): fc.Arbitrary<
  Record<
    Constraints extends { key: fc.Arbitrary<infer Key> } ? Key : string,
    Constraints extends { values: fc.Arbitrary<infer Value>[] }
      ? Value
      : unknown
  >
>

type ObjectConstraints = {
  key?: fc.Arbitrary<string>
  values?: fc.Arbitrary<unknown>[]
}

const a = object({})
//=> fc.Arbitrary<Record<string, unknown>>

const b = object({ key: fc.constantFrom<`a` | `b` | `c`>(`a`, `b`, `c`) })
//=> fc.Arbitrary<Record<"a" | "b" | "c", unknown>>

const c = object({ values: [fc.string(), fc.double()] })
//=> fc.Arbitrary<Record<string, string | number>>
```

**Comment by dubzzz** [comment]
Seems to be a good idea 💡

**Comment by dubzzz** [comment]
Actually the:

```ts
    Constraints extends { values: fc.Arbitrary<infer Value>[] }
      ? Value
      : unknown
```

Is not correct. It might produce more than just Value. I'll need to dig further for that second part.

**Comment by TomerAberbach** [comment]
Hmmm, do you have an example where that's not true? I can't think of one

**Comment by dubzzz** [comment]
Sorry for the late reply, I think I was thinking about things such as:

```ts
fc.object({ values: fc.nat(), withSet: true }) // type Ob = { [k: string]: number | Set<Ob> | Ob }
fc.object({ values: fc.nat() }) // type Ob = { [k: string]: number | Ob }
fc.object({ values: fc.nat(), withTypedArray: true }) // type Ob = { [k: string]: number | TypedArray | Ob }
```

**Comment by TomerAberbach** [comment]
Makes sense. [Something like this would work](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggTgIwJbDgQziAPAFQHxQC8UA3sAFw4C+UAULQCYQDGANhtAGYCuAds8CQB7XlCEIAVi2BYAwiIDOqNEl7AFUCAA9gEXgw0B5SdPm8l6VerwAKWlCjNFyqwvJQzFlWoUAaWgCU7vDIyphYAEoQCkKsAG4QxlICcs6WPnh49KCQUEmmad7qRKT2UADWECAA-MGIKOjhFqoA5lkOcWis3NG1sPVh2HzlvEIA7rx4ANoAumVjKAAWAMoQwH0IQrEQaLxlAPT7msDMAHS0VNng0MtIALZgrEic2PglJFMA0lU411CqFSqQk4UBwM3cOC+P2uMxoADJSFQANz0bRgIRwYBQHLQACiOj0BiwZUMYEEigAjJoCfoNFEnHAGFhmrwWr4oMNRhM8P4HKTyeYAEzU3S0qD0jFMllsjm8EbjSb+AjEAAKSGY5Sw-OE5gp7MqIGBeTJOoUVIRBqN2sUgoI2lFBmNAoUgrK1WxcB6ZXcnC6CggV1yURi8USJhSnhcPhFhKM4eAkfS1hKtweTxekRYkuJDkTRQ09tjpEBIDqoUa2FUnAgcCg3xABBo7vrUHc0t5UAAPlAbHnXDGxSQoJ1utEyw0MJXeNXawA1Lo9aYzKBNqDz0et2XyiYBMrdmz4h0KVLmKPqdlDhbAFZrPq+1j+ruypicVQQBgru00x2oHpQd28BACS1u4qwyMGsQJPkEaFK4mS7g4hzHGctCZPQTBsBwjjOCW46DFgAAGaAEU+BEICR3YEcwBFZBh7BwNATinsOC5jlAUwhBOTSoK0PL9OWk5YLw3B3AgNZ4HMtBMUoUBoCU4jJMANgkFQCGIUcAB61T0NJWIIPJ8bKSWK5qVASFaTpOHMAZilGSOPQaKpZRmZp2lSThH7EAp0h2axfhQFeN4UB6f5OQ4LlQBZOJQAAIiUOJGh+hZipx+FVjWoIEO6kKWiCYKboBwEHK5lnMdAXmGUO9nROygVgT6frQGF6mRdpulQCCFW2VVfm1Us9XPhAr6AR+zURRZtBIQA6tADAiAA5Fi27-CCBF1WsJFIBov7QBiHWNVAiw1hA5ztS0Nk+T1o7+etwX3o+aAaJs2y7CZznmW1OGLBdAi+ddfXXgN93QI9UDPawOyiN2fAvm+o2mR9QA)
